include/dlaf/matrix/hdf5.h

```diff
@@ -0,0 +1,274 @@
+      };
+      const hsize_t file_offsets[3] = {
+          to_sizet(tl.col()),
+          to_sizet(tl.row()),
+          0,
+      };
+      ds_file.selectHyperslab(H5S_SELECT_SET, file_counts, file_offsets);
+
+      const hsize_t memory_dims[3] = {
+          to_sizet(tile.size().cols()),
+          to_sizet(tile.ld()),
+          internal::hdf5_datatype<T>::dims,
+      };
+      H5::DataSpace ds_memory(3, memory_dims);
+
+      const hsize_t memory_counts[3] = {
+          to_sizet(tile.size().cols()),
+          to_sizet(tile.size().rows()),
+          internal::hdf5_datatype<T>::dims,
+      };
+      const hsize_t memory_offsets[3] = {0, 0, 0};
+      ds_memory.selectHyperslab(H5S_SELECT_SET, memory_counts, memory_offsets);
+
+      dataset_op(tile.ptr(), internal::hdf5_datatype<T>::type, ds_memory, ds_file);
+    };
+
+    tt::sync_wait(get_tile(ij) | di::transform(di::Policy<Backend::MC>(), std::move(map_ds_tile)));
+  }
+}
+
+template <class T>
+void from_dataset(const H5::DataSet& dataset, dlaf::Matrix<T, Device::CPU>& matrix) {
+  auto tile_rw = [&matrix](LocalTileIndex ij) { return matrix.readwrite(ij); };
+  auto dataset_read = [&dataset](auto&&... args) {
+    dataset.read(std::forward<decltype(args)>(args)...);
+  };
+  for_each_local_map_with_dataset<T>(matrix.distribution(), dataset, std::move(tile_rw),
+                                     std::move(dataset_read));
+}
+
+template <class T>
+void to_dataset(dlaf::Matrix<const T, Device::CPU>& matrix, const H5::DataSet& dataset) {
+  auto tile_ro = [&matrix](LocalTileIndex ij) { return matrix.read(ij); };
+  auto dataset_write = [&dataset](auto&&... args) {
+    dataset.write(std::forward<decltype(args)>(args)...);
+  };
+  for_each_local_map_with_dataset<T>(matrix.distribution(), dataset, std::move(tile_ro),
+                                     std::move(dataset_write));
+}
+
+}
+
+class FileHDF5 final {
+public:
+  /// File access modes:
+  /// - READONLY, the file will be opened (if should already exist)
+  /// - READWRITE, the file will be created (if should not exist)
+  enum class FileMode {
+    READONLY,
+    READWRITE,
+  };
+
+  /// Create/open a local file.
+  ///
+  /// @p filepath filepath where the file to be opened is or where it will be created
+  /// @p mode file access mode (see FileMode for more details)
+  ///
+  /// @post if mode == READWRITE, being a local file it will not support parallel-write
+  FileHDF5(const std::string& filepath, const FileMode& mode) {
+    file_ = H5::H5File(filepath, mode2flags(mode));
+  }
+
+  /// Create a file that, besides read, supports writing in parallel from different ranks.
+  ///
+  /// @p comm Communicator grouping all ranks that will be able to write in parallel to the file
+  /// @p filepath filepath where the file will be created
+  ///
+  /// @pre @p filepath should not exist
+  /// @post file created will support parallel-write
+  FileHDF5(comm::Communicator comm, const std::string& filepath) {
```

Comment:
`filepath` always first?

Reply:
In general we use grid as first argument.